File: misc/leetcode/cpp/next-greater-element-i.cpp
/* coding:utf-8
 * Copyright (C) dirlt
 */

#include <map>
#include <set>
#include <vector>
using namespace std;

class Solution {
   public:
    vector<int> nextGreaterElement(vector<int>& findNums, vector<int>& nums) {
        vector<int> ans(findNums.size());
        map<int, int> finds;

        for (int i = 0; i < findNums.size(); i++) {
            finds[findNums[i]] = i;
        }
        for (int i = nums.size() - 1; i >= 0; i--) {
            int x = nums[i];
            auto it = finds.find(x);
            if (it == finds.end()) continue;
            int xpos = it->second;

            int value = -1;
            for (int j = i + 1; j < nums.size(); j++) {
                if (nums[j] > x) {
                    value = nums[j];
                    break;
                }
            }
            ans[xpos] = value;
        }
        return ans;
    }
};

File: leetcode/next-greater-element-i.py
#!/usr/bin/env python
# coding:utf-8
# Copyright (C) dirlt

from typing import List


class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        st = []
        values = {}
        while nums2:
            x = nums2.pop()
            while st and st[-1] < x:
                st.pop()
            y = st[-1] if st else -1
            values[x] = y
            st.append(x)

        ans = []
        for x in nums1:
            ans.append(values[x])

        return ans


if __name__ == '__main__':
    pass
